src/ratelimit/RateLimitController.ts

```diff
@@ -0,0 +1,187 @@
+  type: `${typeof name}:stateChange`;
+  payload: [RateLimitState<ApiType>, Patch[]];
+};
+
+export type GetRateLimitState<ApiType extends string> = {
+  type: `${typeof name}:getState`;
+  handler: () => RateLimitState<ApiType>;
+};
+
+export type CallApi<ApiType extends string> = {
+  type: `${typeof name}:call`;
+  handler: RateLimitController<ApiType>['call'];
+};
+
+export type ControllerActions<ApiType extends string> =
+  | GetRateLimitState<ApiType>
+  | CallApi<ApiType>;
+
+export type RateLimitMessenger<
+  ApiType extends string
+> = RestrictedControllerMessenger<
+  typeof name,
+  ControllerActions<ApiType>,
+  RateLimitStateChange<ApiType>,
+  never,
+  never
+>;
+
+const metadata = {
+  requests: { persist: false, anonymous: false },
+};
+
+/**
+ * Controller that handles showing notifications to the user and rate limiting origins
+ */
+export class RateLimitController<ApiType extends string> extends BaseController<
+  typeof name,
+  RateLimitState<ApiType>,
+  RateLimitMessenger<ApiType>
+> {
+  private implementations;
+
+  private rateLimitTimeout;
+
+  private rateLimitCount;
+
+  /**
+   * Creates a RateLimitController instance.
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - A reference to the messaging system.
+   * @param options.state - Initial state to set on this controller.
+   * @param options.implementations - Mapping from API type to API implementation
+   * @param options.rateLimitTimeout - The time window in which the rate limit is applied
+   * @param options.rateLimitCount - The amount of notifications an origin can show in the rate limit time window
+   */
+  constructor({
+    // @todo Pick some sane defaults for this
+    rateLimitTimeout = 5000,
+    rateLimitCount = 3,
+    messenger,
+    state,
+    implementations,
+  }: {
+    rateLimitTimeout?: number;
+    rateLimitCount?: number;
+    messenger: RateLimitMessenger<ApiType>;
+    state?: Partial<RateLimitState<ApiType>>;
+    implementations: Record<ApiType, (...args: unknown[]) => unknown>;
+  }) {
+    const defaultState = {
+      requests: Object.keys(implementations).reduce(
+        (acc, key) => ({ ...acc, [key]: {} }),
+        {} as Record<ApiType, Record<string, number>>,
+      ),
+    };
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.implementations = implementations;
+    this.rateLimitTimeout = rateLimitTimeout;
+    this.rateLimitCount = rateLimitCount;
+
+    this.messagingSystem.registerActionHandler(
+      `${name}:call` as const,
+      (origin: string, args: CallArgs<ApiType>) => this.call(origin, args),
+    );
+  }
+
+  /**
+   * Shows a notification if origin is not rate-limited.
+   *
+   * @param origin - The origin trying to send a notification
+   * @param _args - Notification arguments, containing the notification message etc.
+   * @returns False if rate-limited, true if not
+   */
```

Comment:
```suggestion
  /**
   * Calls an API if the requesting origin is not rate-limited.
   *
   * @param origin - The requesting origin.
   * @param _args - Arguments for the API call.
   * @returns `false` if rate-limited, and `true` otherwise.
   */
```